Support variable flash and RAM start addresses and sizes.

david-perez opened the pull request:
Previously, only variable flash start address was supported via the
FLASH_START cmake variable (see commit 2d9a530).

Commit 160991c introduced a partial regression in accidentally removing
support for variable flash start addresses. This commit fixes that,
while in addition introducing support for variable flash size, RAM start
address and RAM size.

Fixes #9.

david-perez committed:
Support variable flash and RAM start addresses and sizes.

Previously, only variable flash start address was supported via the
FLASH_START cmake variable (see commit 2d9a530).

Commit 160991c introduced a partial regression in accidentally removing
support for variable flash start addresses. This commit fixes that,
while in addition introducing support for variable flash size, RAM start
address and RAM size.

ryankurte reviewed (changes requested):
hey thanks for the PR, looks excellent! a couple of thoughts on the way

ryankurte commented on CMakeLists.txt:
```diff
@@ -59,6 +59,31 @@ set(SOURCES
 
 # Generate executable and link
 add_executable(${TARGET} ${SOURCES})
+
```
i think it makes sense to to have these definitions in `efm32-base.cmake` rather than here, to separate use of the tools from the supporting logic and because it's common to any `CMakeLists.txt`?

david-perez replied:
I agree, but [`target_link_options`](https://cmake.org/cmake/help/git-stage/command/target_link_options.html) can only be used after we've defined the target via `add_executable`.

ryankurte replied:
hmm, fair. what if we put the link options into a cmake function in `efm32-base.cmake` then add a call to `CMakeLists.txt` after `add_executable`? (something like `efm32_configure_linker(${TARGET})`)

david-perez replied:
Something like this?

ryankurte replied:
look perfect, thanks ^_^

ryankurte commented on README.md:
```diff
@@ -58,6 +58,8 @@ Once you have integrated this project with your project, you can build in the st
 4. `make` to execute the build
 5. `make flash` to flash to a device
 
+When configuring the build, you may want to define using the `-D` flag the variables `DEVICE`, `FLASH_ORIGIN`, `FLASH_LENGTH`, `RAM_ORIGIN` and `RAM_LENGTH`. If not defined, the build will use the default memory adresses and sizes for the FLASH and RAM regions defined in the Silicon Labs provided linker script for your device (these are usually `0x00000000` and `0x20000000` for flash start and RA ...
```
Possible change, depending on your application it's useful to configure this from the command line or in your own `CMakeLists.txt`

```suggestion
When configuring the build, you may want to define (either in `CMakeLists.txt`or using the `-D` flag) the variables `DEVICE`, `FLASH_ORIGIN`, `FLASH_LENGTH`, `RAM_ORIGIN` and `RAM_LENGTH`. If not defined, the build will use the default memory addresses and sizes for the FLASH and RAM regions defined in the Silicon Labs provided linker script for your device (these are usually `0x00000000` and `0x20000000` for flash start and RAM start, respectively).
```

ryankurte commented on device/EFM32G/Source/GCC/efm32g.ld:
```diff
@@ -30,8 +30,8 @@
 {
-  FLASH (rx) : ORIGIN = 0x00000000, LENGTH = 131072
-  RAM (rwx)  : ORIGIN = 0x20000000, LENGTH = 16384
+  FLASH (rx) : ORIGIN = DEFINED(flash_origin) ? flash_origin : 0x00000000, LENGTH = DEFINED(flash_length) ? flash_length : 131072
```
ahh this is great!

david-perez committed:
Update README.md

Co-Authored-By: Ryan <ryankurte@users.noreply.github.com>

david-perez commented:
For future reference, I used `find . -type f -name "*.ld" -exec sed -i -r '/MEMORY/ { n; n; s/(ORIGIN = )(.*), (LENGTH = )(.*)/\1DEFINED(flash_origin) ? flash_origin : \2, \3DEFINED(flash_length) ? flash_length : \4/; n; s/(ORIGIN = )(.*), (LENGTH = )(.*)/\1DEFINED(ram_origin) ? ram_origin : \2, \3DEFINED(ram_length) ? ram_length : \4/; }' {} +` to change the linker scripts from the `device/` folder.

david-perez committed:
Move definition of linker addresses to function in efm32base.cmake

david-perez committed:
Merge branch 'master' of https://github.com/david-perez/efm32-base

david-perez committed:
Fix assertion in linker script that checks if flash overflowed

david-perez commented:
I fixed the last assertion in the linker script that checks that flash memory doesn't overflow, since now programs may not necessarily start at `0x0`.

For future reference, this time I used `find . -type f -name "*.ld" -exec sed -i -r 's/(>= \(__etext)/\1 - ORIGIN\(FLASH\)/' {} +` from the `device/` folder.

ryankurte commented:
looks awesome, thanks!

are you able to refactor against master to sort out those conflicts, then we can merge this in?

david-perez committed:
Merge branch 'master' into master

david-perez committed:
Fix call to efm32_configure_linker_addresses

david-perez committed:
Fix linker scripts of new devices

Commit 08571391753d6f7376173a8e910422d5a12d14c0 added lots of devices
under the device/ folder. This commit fixes their linker scripts to
support variable flash and RAM start addresses and sizes.

david-perez commented:
That should do it.
